extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergTableAdapter.java

```diff
@@ -228,231 +229,164 @@ public synchronized Optional<Schema> schema(final int schemaId) {
-        final Snapshot tableSnapshot =
-                snapshot(snapshotId).orElseThrow(() -> new IllegalArgumentException(
-                        "Snapshot with id " + snapshotId + " not found for table " + tableIdentifier));
-
-        // Load the table from the catalog.
-        return definition(tableSnapshot, instructions);
+    private static final class SpecAndSchema {
+        private final Schema schema;
+        private final PartitionSpec partitionSpec;
+        private final IcebergReadInstructions readInstructions;
+
+        private SpecAndSchema(
+                @NotNull final Schema schema,
+                @NotNull final PartitionSpec partitionSpec,
+                @NotNull final IcebergReadInstructions readInstructions) {
+            this.schema = schema;
+            this.partitionSpec = partitionSpec;
+            this.readInstructions = readInstructions;
+        }
     }
 
     /**
-     * Return {@link TableDefinition table definition} for the Iceberg table and snapshot, with optional instructions
-     * for customizations while reading.
-     *
-     * @param tableSnapshot The snapshot to load
-     * @param instructions The instructions for customizations while reading (or null for default instructions)
-     * @return The table definition
+     * Retrieve the schema and partition spec for the table based on the provided read instructions. Also, populate the
+     * read instructions with the requested snapshot, or the latest snapshot if none is requested.
      */
-    public TableDefinition definition(
-            @Nullable final Snapshot tableSnapshot,
-            @Nullable final IcebergInstructions instructions) {
-
+    private SpecAndSchema getSpecAndSchema(@NotNull IcebergReadInstructions readInstructions) {
+        final Snapshot snapshot;
         final Schema schema;
-        final org.apache.iceberg.PartitionSpec partitionSpec;
+        final PartitionSpec partitionSpec;
 
-        if (tableSnapshot == null) {
+        final Snapshot snapshotFromInstructions = getSnapshot(readInstructions);
+        if (snapshotFromInstructions == null) {
             synchronized (this) {
                 // Refresh only once and record the current schema and partition spec.
                 refresh();
+                snapshot = table.currentSnapshot();
                 schema = table.schema();
                 partitionSpec = table.spec();
             }
+            if (snapshot != null) {
+                // Update the read instructions with the snapshot.
+                readInstructions = readInstructions.withSnapshot(snapshot);
+            }
         } else {
             // Use the schema from the snapshot
-            schema = schema(tableSnapshot.schemaId()).get();
+            snapshot = snapshotFromInstructions;
+            schema = schema(snapshot.schemaId()).get();
             partitionSpec = table.spec();
+            readInstructions = readInstructions.withSnapshot(snapshot);
```

Comment:
For readability and correctness, I'd prefer to see input parameter `readInstructions` as `final` and assign a `final IcebergReadInstructions instructions` to the appropriate (potentially modifed) instructions object.